### PR TITLE
Enable CORS for GraphQL APIs

### PR DESCRIPTION
**Pull request by ayyazdaniaryan**
Enabling CORS only for specified websites.

### CODE REVIEW

File: assembl/models/preferences.py [hunk 1]
```diff
@@ -289,8 +290,12 @@ def validate_single_value(self, key, value, pref_data, data_type):
...
             elif data_type == "url":
                 from urlparse import urlparse
-                assert urlparse(value).scheme in (
-                    'http', 'https'), "Not a HTTP URL"
+                result = urlparse(value).scheme in (
+                    'http', 'https')
+                if not result:
+                    # Check if it's an IP address
+                    result = is_valid_ipv4_address(value)
+                assert result, "Not a valid address"
```
Comment (cedricmessiant) [inline assembl/models/preferences.py]: maybe you could keep the old `assert` to have a different error message if it's not a http url?

Comment (ayyazdaniaryan) [reply assembl/models/preferences.py]: Sure, can look further into it.

File: assembl/models/preferences.py [hunk 2]
```diff
@@ -965,7 +971,20 @@ def get_preference_data_list(self):
...
-        }
+        },
+
+        # Valid CORS paths
+        {
+            "id": "graphql_valid_cors",
+            "name": _("Valid CORS paths for GraphQL API calls"),
+            "value_type": "list_of_url",
+            "show_in_preferences": True,
+            "description": _("A list of valid URLs or IP addresses that are allowed to make CORS api calls to the GraphQL API"),
```
Comment (maparent) [inline assembl/models/preferences.py]: domain names or IP addresses

File: assembl/graphql/view.py [hunk 3]
```diff
@@ -79,4 +99,7 @@ def graphql_api(request):
     # doesn't play nice with multiprocesses or multithreading and sqlalchemy session.
     # We monkey patch get_graphql_params for logging instead.
     solver = graphql_wsgi_wrapper(Schema)  # , middleware=[LoggingMiddleware()])
-    return solver(request)
+    response = solver(request)
+    if has_cors:
+        response.headerlist.extend(cors_headers)
+    return response
```
Comment (vincentfretin) [inline assembl/graphql/view.py]: LGTM

File: assembl/models/preferences.py [hunk 4]
```diff
@@ -289,8 +290,12 @@ def validate_single_value(self, key, value, pref_data, data_type):
                     "value not allowed: " + value)
             elif data_type == "url":
```
[the remark this answers is not in the document]
Comment (ayyazdaniaryan) [reply assembl/models/preferences.py]: Yeah, I didn't want to change domain, as it's used for addresses who have an MX DNS record. I created a new `address` type, and check against IPv4, IPv6, and URL type.

File: assembl/models/preferences.py [hunk 5]
```diff
@@ -288,14 +288,29 @@ def validate_single_value(self, key, value, pref_data, data_type):
...
+                        condition = True
+                        break
+                    elif is_valid_ipv4_address(value):
+                        condition = True
+                        break
+                    elif is_valid_ipv6_address(value):
+                        condition = True
+                        break
+                    else:
+                        condition = bool(urlparse(value).scheme)
```
Comment (maparent) [inline assembl/models/preferences.py]: Same comment: Address and URL are distinct types. I'm fine if you accept a domain at this line, not a URL.

**Pull request closed without merging**